File: message-queue/golang/primitives.go
package golang

import "time"
import "math"

type WorkerResult struct {
	WorkerId      int
	MessagesTotal int
	DurationNs    time.Duration

	MessagesPerSecond float64
	DurationSeconds   float64
}

func NewWorkerResult(workerId int, messagesTotal int, durationNs time.Duration) *WorkerResult {
	r := new(WorkerResult)

	r.WorkerId = workerId
	r.MessagesTotal = messagesTotal
	r.DurationNs = durationNs

	r.DurationSeconds = float64(durationNs) * math.Pow10(-9)
	r.MessagesPerSecond = float64(messagesTotal) / r.DurationSeconds

	return r
}

type Results struct {
	Workers       []*WorkerResult
	MessagesTotal int
	DurationNs    time.Duration

	MessagesPerSecond float64
	DurationSeconds   float64
}

func NewResults() *Results {
	r := new(Results)
	return r
}

func (rs *Results) Add(r *WorkerResult) {
	rs.Workers = append(rs.Workers, r)
	rs.MessagesTotal += r.MessagesTotal
	rs.DurationNs += r.DurationNs

	rs.DurationSeconds = float64(rs.DurationNs) * math.Pow10(-9)
	rs.MessagesPerSecond = float64(rs.MessagesTotal) / r.DurationSeconds
}

File: message-queue/cpp/src/primitives.cpp
#ifndef PRIMITIVES_CPP
#define PRIMITIVES_CPP
#include "./primitives.hpp"

WorkerResult::WorkerResult(int WorkerId, int MessagesTotal, WorkDuration Duration)
: WorkerId(WorkerId)
, MessagesTotal(MessagesTotal)
, Duration(Duration)
, DurationSeconds(Duration.count())
, MessagesPerSecond(MessagesTotal / DurationSeconds)
{
}

void Results::Add(WorkerResult wr) {
  Workers.push_back(wr);
  MessagesTotal += wr.MessagesTotal;
  Duration += wr.Duration;

  DurationSeconds = Duration.count();
  MessagesPerSecond = Workers.size() * MessagesTotal / DurationSeconds;
}

void Results::Print(logger& log) const {
  for(auto& wr : Workers) {
    log.info(string(wr));
  }

  auto s = fmt::format("Total: {}", MessagesTotal);
  log.info(s);
  s = fmt::format("Total mps: {}", MessagesPerSecond);
  log.info(s);
}

map<string,string> SampleDesc::Map() const {
  map<string,string> xs = {
    {"n_workers", to_string(n_workers)},
    {"algorithm", algorithm},
    {"mq_system", mq_system},
  };
  return xs;
}

#endif
